Improved User Guide for OVE (#13)
ove/ove-docs: docs/USAGE.md

There are several steps that must be performed in order to use OVE to control a display.

## Setting up OVE

Before using OVE, you will need to install [**OVE Core**](https://github.com/ove/ove), any [**OVE Apps**](https://github.com/ove/ove-apps), and any [**OVE Services**](https://github.com/ove/ove-services) that you intend to use. Installation guidelines can be found in the [OVE Installation Guide](./INSTALLATION.md).

You will also need to modify the [``ove/packages/ove-core/src/client/Spaces.json``](https://github.com/ove/ove/blob/master/packages/ove-core/src/client/Spaces.json) file to define a new ``space`` in OVE core.
## Setting up OVE

> Note that all services need to hosted somewhere that is accessible from the computers connected to the monitors that will be used in the display.

Reviewer: I think that it would still be useful to make a statement like this (yes, it is mentioned in the install guide, but I think it is worth repeating).

Author: @jamesscottbrown, I see few issues in doing that. The installation guide explains how to do this very precisely. If you want to duplicate the same, you will have to include more than just this line. Also, something like `need to be hosted somewhere` does not make sense. Also, even if we ignore that simply `hosting the services somewhere where they are accessible from the computers connected to the monitors` will also not work, as the configuration is what matters. This line is half baked and that is why I removed it. Hope we can agree on this?

Reviewer: The problem is that at the time that OVE is installed it may not be known how the machines that will be running the OVE clients will be connected to the network.

(As a real-world example of this, we deployed OVE to a server with the intention of displaying things on the GDO and DODev, but then tried to display things on MultiTaction displays that we didn't know we'd receive at the time we did the installation.)

This document describes how to "create" (instantiate?) an OVE space. It is at *this time*, not only the *installation time* that OVE was installed on a server, that you need to ensure that the server and client browsers can access each other.

I agree that the previous phrasing was imperfect. How about:

> You must ensure that the OVE core server, and all OVE services, are accessible from the computers connected to the monitors that will be used in the display.

> You will need to modify the [`Spaces.json`](https://github.com/ove/ove/blob/master/packages/ove-core/src/client/Spaces.json) file and introduce a new `space` in OVE, before it can be used in a production setting.

Reviewer: ```suggestion
You will need to modify the [`Spaces.json`](https://github.com/ove/ove/blob/master/packages/ove-core/src/client/Spaces.json) file and introduce a new `space` in OVE, before it can be used.
```

If the space isn't defined it can't be used *at all*, whether for testing/experimentation or in production.

## Launching browsers

> In order to use OVE, you will need to open the URLs of the `clients` in a web browser. Each URL has the form: `http://OVE_CORE_HOST:PORT/view.html?oveViewId={SPACE}-{ID}`. The values for `OVE_CORE_HOST` and `PORT` must be set as explained in the [OVE Installation Guide](./INSTALLATION.md#running-ove). The value for `SPACE` would be the name of the `space` used, such as `LocalNine`, `LocalFour`, or a name of a new `space` that has been defined in the [`Spaces.json`](https://github.com/ove/ove/blob/master/packages/ove-core/src/client/Spaces.json) file. The value for `ID` is the index of the `client` associated with the browser window, in the definition of the `space` in [`Spaces.json`](https://github.com/ove/ove/blob/master/packages/ove-core/src/client/Spaces.json) file.

Reviewer: "open the URLs of **the `clients`** in a web browser" -> "open the URL of **each `client`** in a **separate** web browser"

```suggestion
In order to use OVE, you will need to open the URL of each `client` in a separate web browser. Each URL has the form: `http://OVE_CORE_HOST:PORT/view.html?oveViewId={SPACE}-{ID}`. The values for `OVE_CORE_HOST` and `PORT` must be set as explained in the [OVE Installation Guide](./INSTALLATION.md#running-ove). The value for `SPACE` would be the name of the `space` used, such as `LocalNine`, `LocalFour`, or a name of a new `space` that has been defined in the [`Spaces.json`](https://github.com/ove/ove/blob/master/packages/ove-core/src/client/Spaces.json) file. The value for `ID` is the index of the `client` associated with the browser window, in the definition of the `space` in [`Spaces.json`](https://github.com/ove/ove/blob/master/packages/ove-core/src/client/Spaces.json) file.
```

> It is recommended to use OVE with Google Chrome, as this is the web browser used for development and in production at the [Data Science Institute](http://www.imperial.ac.uk/data-science/). However, it should also be compatible with other modern web browsers: if you encounter any browser-specific bugs please [report them as an Issue](https://github.com/ove/ove-apps/issues).

Reviewer: ```suggestion
It is recommended to use OVE with Google Chrome, as this is the web browser used for development and in production at the [Data Science Institute](http://www.imperial.ac.uk/data-science/). However, OVE should also be compatible with other modern web browsers: if you encounter any browser-specific bugs please [report them as an Issue](https://github.com/ove/ove-apps/issues).
```

> To test OVE functionality, you can simply open four web browser windows (or tabs) with the following URLs to create the `LocalFour` space:

Reviewer: Is this *creating* the space?

Author: Good question: we need a better word here. How about `load`?

Author: I have rewritten as `corresponding to`. It means neither create or load since `space` has a meaning.

```text
http://OVE_CORE_HOST:PORT/view.html?oveViewId=LocalFour-0
http://OVE_CORE_HOST:PORT/view.html?oveViewId=LocalFour-1
http://OVE_CORE_HOST:PORT/view.html?oveViewId=LocalFour-2
http://OVE_CORE_HOST:PORT/view.html?oveViewId=LocalFour-3
```

However, in a much larger OVE installation with many monitors, opening browser windows may be sufficiently time-consuming for automation to be worthwhile.

> In a Windows environment, a [PowerShell](https://docs.microsoft.com/en-us/powershell/scripting/powershell-scripting) script can be used to launch full-screen browsers with the correct URLs. In a Linux or MacOS environment, a [Bash](https://linux.die.net/man/1/bash) script can be used to launch full-screen browsers with the correct URLs. In a MacOS environment, [AppleScript](https://developer.apple.com/library/archive/documentation/AppleScript/Conceptual/AppleScriptLangGuide/introduction/ASLR_intro.html) can be used as an alternative to [Bash](https://linux.die.net/man/1/bash).

Reviewer: ```suggestion
On Windows, a [PowerShell](https://docs.microsoft.com/en-us/powershell/scripting/powershell-scripting) script can be used to launch full-screen browsers with the correct URLs. On Linux, a [Bash](https://linux.die.net/man/1/bash) script can be used. On MacOS, either [AppleScript](https://developer.apple.com/library/archive/documentation/AppleScript/Conceptual/AppleScriptLangGuide/introduction/ASLR_intro.html) or [Bash](https://linux.die.net/man/1/bash) could be used.
```

Author: Actually speaking AppleScript is not the first preference on a Mac. I will therefore reverse the order and include this.

## Launching OVE Apps

OVE provides three different ways of launching apps into an OVE environment:

* Using [**OVE Editor**](https://github.com/ove/ove-editor)
* Using [**OVE SDKs**](https://github.com/ove/ove-sdks)
* Using [**OVE Core**](https://github.com/ove/ove) APIs

The [**OVE Editor**](https://github.com/ove/ove-editor) is designed to be used by the most basic users of OVE. The [**OVE Core**](https://github.com/ove/ove) APIs are intended to be used by the most advanced users of OVE.

Given below are instructions on how to load sample content using the [Image App](https://github.com/ove/ove-apps/tree/master/packages/ove-app-images). A complete list of all apps and similar instructions on how to use them can be found in the [**OVE Apps**](https://github.com/ove/ove-apps) repository.

### Launching OVE Apps from the OVE Editor

The [**OVE Editor**](https://github.com/ove/ove-editor) is still work in progress, and is not currently available as an option.

### Launching OVE Apps using the Python Client Library

> The [Python Client Library](https://github.com/ove/ove-sdks/tree/master/python) (installation instructions available on GitHub repository) is one of the SDKs provided by OVE. To launch the [Image App](https://github.com/ove/ove-apps/tree/master/packages/ove-app-images) with a sample image, in the `LocalNine` space, run:

Reviewer: I don't think that "(installation instructions available on GitHub repository)" is necessary: you see the instructions as soon as you follow the link.

Author: We have had situations where people did not realise they had to click the link to figure that out. I will therefore leave this as-is.

Reviewer: Ok - it would perhaps be more elegant to write "The [Python Client Library](https://github.com/ove/ove-sdks/tree/master/python) ([installation instructions](https://github.com/ove/ove-sdks/blob/master/python/README.md))"

i.e.,

     "The [Python Client Library](https://github.com/ove/ove-sdks/tree/master/python) ([installation instructions](https://github.com/ove/ove-sdks/blob/master/python/README.md))"

Author: In terms of the installation instructions, the links https://github.com/ove/ove-sdks/tree/master/python and https://github.com/ove/ove-sdks/blob/master/python/README.md would be idempotent as GitHub expands the contents of the README.md when viewing a directory. 

As a resolution, I will make it `The [Python Client Library](https://github.com/ove/ove-sdks/tree/master/python) (installation instructions available on [GitHub repository](https://github.com/ove/ove-sdks/blob/master/python/README.md#installation))`. This will mean those who read a printed/pdf version of the docs benefit as much as those who read the web version.

```python
from ove.config import local_space as space

space.enable_online_mode()

space.delete_sections()

image = space.add_section(w=4320, h=2424, x=0, y=0, app_type='images')
image.set_url("https://farm4.staticflickr.com/3107/2431422903_632ce51b56_o_d.jpg", "shelley")
```

### Launching OVE Apps using OVE Core APIs

OVE provides a number of useful APIs that can be used to launch applications and load content. The complete list of APIs provided by [**OVE Core**](https://github.com/ove/ove) is documented at: `http://OVE_CORE_HOST:PORT/api-docs`.

> OVE core, all apps, and all services use [Swagger](https://swagger.io/solutions/api-documentation/) to document these APIs, which makes it possible to directly invoke them from within the documentation page (`http://OVE_CORE_HOST:PORT/api-docs`) opened on a web browser. A standalone tool such as [curl](https://curl.haxx.se/docs/manpage.html) can be used as an alternative to the [Swagger](https://swagger.io/solutions/api-documentation/) UI.

Reviewer: ```suggestion
The APIs for OVE core, all apps, and all services are documented using [Swagger](https://swagger.io/solutions/api-documentation/), and it possible to directly invoke them from within the documentation page (`http://OVE_CORE_HOST:PORT/api-docs`) using a web browser. Alternatively, a standalone tool such as [curl](https://curl.haxx.se/docs/manpage.html) can be used.
```

An image can be loaded into the `LocalFour` space using the OVE APIs, by running the following commands using [curl](https://curl.haxx.se/docs/manpage.html).

Linux/Mac:

```sh
curl --header "Content-Type: application/json" --request DELETE http://OVE_CORE_HOST:PORT/section
curl --header "Content-Type: application/json" --request POST --data '{"app": {"url": "http://OVE_APP_IMAGES_HOST:PORT","states": {"load": {"tileSources": "https://openseadragon.github.io/example-images/highsmith/highsmith.dzi"}}}, "space": "LocalFour", "h": 500, "w": 500, "y": 0, "x": 0}' http://OVE_CORE_HOST:PORT/section
```

Windows:

```sh
curl --header "Content-Type: application/json" --request DELETE http://OVE_CORE_HOST:PORT/section
curl --header "Content-Type: application/json" --request POST --data "{\"app\": {\"url\": \"http://OVE_APP_IMAGES_HOST:PORT\", \"states\": {\"load\": {\"tileSources\": \"https://openseadragon.github.io/example-images/highsmith/highsmith.dzi\"}}}, \"space\": \"LocalFour\", \"h\": 500, \"w\": 500, \"y\": 0, \"x\": 0}" http://OVE_CORE_HOST:PORT/section
```

> These commands clear all sections on OVE, and creates a new section with an instance of the [Image App](https://github.com/ove/ove-apps/tree/master/packages/ove-app-images) running within it. In this example we are loading an image with a [Deep Zoom](https://docs.microsoft.com/en-us/previous-versions/windows/silverlight/dotnet-windows-silverlight/cc645077(v=vs.95)) file format.

Reviewer: ```suggestion
These commands clear all sections on OVE, and create a new section containing an instance of the [Image App](https://github.com/ove/ove-apps/tree/master/packages/ove-app-images). In this example we are loading an image in the [Deep Zoom](https://docs.microsoft.com/en-us/previous-versions/windows/silverlight/dotnet-windows-silverlight/cc645077(v=vs.95)) file format.
```

Author: Is it `image with a format` or `image in the format`? You cannot load the same image in a different format, for instance.

Author: Resolved in a much better way.

## Controlling OVE Apps and designing interactive visualisations

> Once an App has been loaded into an OVE `space` it can be controlled using the corresponding controller, that provides app-specific functionality. The controller of the [Image App](https://github.com/ove/ove-apps/tree/master/packages/ove-app-images) supports panning and zooming of images that have been loaded. The controller can be loaded into a separate browser window by accessing the URL `http://OVE_APP_IMAGES_HOST:PORT/control.html?oveSectionId=0`.

Reviewer: ```suggestion
Once an App has been loaded into an OVE `space` it can be controlled using the corresponding controller, which provides app-specific functionality. For example, the controller of the [Image App](https://github.com/ove/ove-apps/tree/master/packages/ove-app-images) supports panning and zooming of images that have been loaded. The controller can be loaded into a separate browser window by accessing the URL `http://OVE_APP_IMAGES_HOST:PORT/control.html?oveSectionId=0`.
```

> A common practice when designing projects with interactive visualisations is to create a custom launcher application that is capable of making API calls. Such applications are usually designed to run on web browsers and invoke [**OVE Core**](https://github.com/ove/ove) API using JavaScript code. These applications eliminate the need of manual intervention and automate the launching and controlling of OVE apps, providing a single-click (or single-touch) experience.

Reviewer: ```suggestion
A common practice when designing projects with interactive visualisations is to create a custom launcher application that is capable of making API calls. Such applications are usually designed to run on web browsers and access the [**OVE Core**](https://github.com/ove/ove) API using JavaScript code. These applications provide a single-click (or single-touch) experience for launching and controlling OVE apps.
```

Author: Not sure if this is acceptable in full. Are you `accessing` an API or `invoking` an API?

Reviewer: I'm not sure about access vs. invoking, but you're definitely not invoking "an API", but rather an *API method*.

Reviewer: Whereas I think *accessing* "an API" might be ok.

Author: Googling reveals that a wide majority of the middleware vendors seem to be using `invoking an API` while `access an API` concerns things like credentials that you need to use in order to `access the API`. So, I will keep it as `invoke an API`.
